Technical Training - Alecc

peroksiid opened the pull request.

Commit:
Update folder structure and add Technical Training link for Odoo 16.0

Commit:
Update estate module version to 18.0.1.0.0 and standardize manifest formatting

Commit:
Rename TestModel to EstateProperty and add description for clarity

Commit:
Add basic fields to EstateProperty model including title, description, and amenities

Commit:
Add security access control for EstateProperty model and update manifest data order

Commit:
Implement estate property views and update manifest to include new view file

Commit:
Add estate menus XML file and update manifest to include it

Commit:
Update view mode in estate property action from 'tree' to 'list'

Commit:
Enhance EstateProperty model by adding default values for date_availability and state fields, and updating selling_price to be non-copyable. Introduce active field for property status management.

Commit:
Expand status options in EstateProperty model by adding 'Offer Received', 'Offer Accepted', 'Sold', and 'Cancelled' to the state field selection.

Commit:
Add list view for estate.property model to display key property details including name, postcode, bedrooms, living area, expected price, selling price, and date availability.

Commit:
Add form view for estate.property model to facilitate detailed property entry, including fields for name, availability date, pricing, description, and various property features.

Commit:
Add search view for estate.property model to enable efficient property searching with quick search fields, filters, and grouping options.

Commit:
Add estate property type model and views, including list and form views, and update estate property model to include property type relation. Enhance manifest and access control for new model.

Commit:
Add estate property tag model, views, and access control. Update estate property model to include many2many relationship for tags and enhance manifest and menu structure to support new functionality.

Commit:
Add estate property offer model, views, and access control. Update estate property model to include one2many relationship for offers and enhance manifest to support new functionality.

Commit:
Change view mode from 'tree' to 'list' for offers in estate property views to improve consistency and clarity in the user interface.

Commit:
Update context for offer_ids field in estate property views to use 'id' instead of 'active_id' for improved data accuracy.

Commit:
Add computed fields for total area and best price in EstateProperty model, and implement validity and deadline fields in EstatePropertyOffer model. Update views to display new fields.

Commit:
Add onchange methods for validity and date_deadline in EstatePropertyOffer model, and implement onchange method for garden in EstateProperty model to automatically set garden area and orientation based on garden presence.

Commit:
Add action methods for accepting and refusing offers in EstatePropertyOffer model, and implement cancellation and sold state actions in EstateProperty model. Update views to include new action buttons for improved user interaction.

Commit:
gitignore: ignore Markdown files (*.md) and stop tracking existing .md files

Commit:
Refactor button attributes in estate property views to simplify visibility conditions for 'Sold' and 'Cancel' actions.

Commit:
Enhance button attributes in estate property views by adding icons for 'Sold' and 'Cancel' actions to improve user interface clarity.

Commit:
Update offer buttons in estate property views to include icons and adjust visibility conditions based on offer status for improved user experience.

Commit:
Add action buttons for accepting and refusing offers in estate property views, with icons and visibility conditions based on offer status to enhance user interaction.

Commit:
Add SQL constraints to ensure positive pricing in EstateProperty, EstatePropertyOffer, EstatePropertyTag, and EstatePropertyType models for data integrity.

Commit:
Add Python-level validation constraints in EstateProperty model to ensure expected and selling prices are positive, and enforce a minimum selling price threshold based on expected price for improved data integrity and user feedback.

Commit:
Enhance estate models by adding new fields for property type and color, updating view definitions for improved data representation and user interaction. Implement ordering for offers and property types to streamline listings.

Commit:
Refactor offer field in estate property views to streamline readonly conditions and enhance user interaction by simplifying the list display for offers.

Commit:
Add action for estate property offers in views to enhance user navigation and interaction by allowing access to offer listings based on property type.

Commit:
Remove unused action definition for estate property offers to streamline view files and improve maintainability.

Commit:
Refactor estate property type model and views by removing unused fields and simplifying the structure to enhance maintainability and user experience.

Commit:
Add property relationship to EstatePropertyType model and update views to display related properties. Introduced a new notebook page for properties in the estate property type form for improved data organization and user interaction.

Commit:
Enhance EstatePropertyOffer and EstateProperty models by adding cascading delete behavior for property offers and ensuring related offers are removed upon property deletion to maintain data integrity.

Commit:
Add ordering and sequence fields to EstatePropertyTag and EstatePropertyType models, updating views accordingly for improved data organization and user interaction.

Commit:
Update estate property type views to enhance list display with conditional decorations for property states, improving visual feedback for users.

Commit:
Update estate property list view to apply conditional decorations for multiple offer states, enhancing visual feedback for users.

Commit:
Remove estate property model file and update estate property list view to include priority field for improved display order.

Commit:
Update estate property list view to change decoration style for 'offer_accepted' state, enhancing visual distinction for users.

Commit:
Update estate property list view to include color options for tags and reference the view ID, improving visual organization and navigation.

Commit:
Update estate property list view to correct decoration attribute for 'offer_accepted' state, enhancing visual clarity for users.

Commit:
Add logic to mark property as having an offer upon creation of EstatePropertyOffer, ensuring state consistency for new offers.

Commit:
Update estate property list view to modify button visibility logic based on offer status, improving user interaction by preventing actions on accepted or refused offers.

Commit:
Add offer management features to EstatePropertyType model and views

- Introduced a One2many relationship for offers in the EstatePropertyType model.
- Added a computed field to count the number of offers associated with each property type.
- Updated the estate_property_offer_views.xml to include an action for managing offers.
- Enhanced estate_property_type_views.xml to display offer count and provide a button for accessing offers, improving user interaction and data visibility.

Commit:
Refactor estate manifest to remove duplicate offer view entry and maintain clarity in view definitions.

Commit:
Fix button context in estate_property_type_views.xml to use 'id' instead of 'active_id', ensuring correct offer filtering in the Offers action.

Commit:
Add property management features for users

- Introduced a One2many relationship in the ResUsers model to link properties assigned to salespersons, filtered by state (New or Offer Received).
- Updated the user form view to display the associated properties, enhancing user experience and property management capabilities.
- Implemented business logic in EstatePropertyOffer to ensure offers are strictly higher than existing ones, maintaining data integrity.

Commit:
Add estate accounting integration module

- Created the estate_account module to link estate properties with accounting functionalities.
- Implemented the EstateProperty model to handle invoice creation upon property sale, including commission and administrative fees.
- Added module manifest with dependencies on estate and account modules.

Commit:
Add kanban view for estate properties and update action view mode

- Introduced a minimal kanban view for estate properties to enhance visual representation and user interaction.
- Updated the estate property action to include kanban in the view mode, allowing users to switch between kanban, list, and form views seamlessly.

Commit:
Enhance kanban view for estate properties with additional fields and improved layout

- Updated the kanban view to group by property type and made records non-draggable for better organization.
- Added fields for expected price, best offer, selling price, and tags to provide more information at a glance.
- Improved the visual layout of the kanban box for better user experience.

Commit:
Refactor estate_property model to enhance validation and action methods

- Rearranged import statements for consistency.
- Added SQL constraints to ensure expected and selling prices are valid.
- Refactored action methods for setting properties as sold or cancelled to use `ensure_one()` for better safety.
- Removed redundant SQL constraints from the model definition, streamlining the code.

peroksiid commented:
@jnc-odoo

jnc-odoo commented on estate/__manifest__.py:
```diff
@@ -1,11 +1,28 @@
+        "views/estate_menus.xml",
+        # "views/estate_property_type_views.xml",
+        # "views/estate_property_tag_views.xml",
+        # "views/estate_property_offer_views.xml",
```
Commented code should be removed

jnc-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,180 @@
+    active = fields.Boolean(string="Active", default=True)
+    state = fields.Selection(
+        selection=[
+            ('new', 'New'),
```
We tend to use either single quote or double quote everywhere to keep a consistency

jnc-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,180 @@
+
+    @api.depends("living_area", "garden_area")
+    def _compute_total_area(self) -> None:
+        for record in self:
```
```suggestion
        for property in self:
```
Usually we use a variable that reflect the model were are in to help the code reader

jnc-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,114 @@
+    def _compute_date_deadline(self) -> None:
+        for offer in self:
+            created = (offer.create_date or fields.Datetime.now())
+            base_date = created.date() if hasattr(created, "date") else date.today()
```
If I am not mistaken: date is not defined ?

jnc-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,114 @@
+    def _inverse_date_deadline(self) -> None:
+        for offer in self:
+            created = (offer.create_date or fields.Datetime.now())
+            base_date = created.date() if hasattr(created, "date") else date.today()
```
```suggestion
            created = (offer.create_date or fields.Datetime.now())
            base_date = created.date()
```
I think that with your first line you would not have an error becasue created should be a datetime in any case ?

jnc-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,114 @@
+        return True
+
+    def action_refuse(self):
+        self.write({"status": "refused"})
```
```suggestion
        self.status = "refused"
```
It is better to let the orm decide when the write should be applied

jnc-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,114 @@
+        # Set property state to Offer Received for affected properties
+        for prop in props:
+            if prop.state in ("new", "offer_received"):
+                prop.state = "offer_received"
```
```suggestion
        props.filtered(lambda p: p.state == "new").state = "offer_receive"
```

jnc-odoo reviewed (commented):
good job !
Only some minor comments

Commit:
Refactor estate models for improved clarity and functionality

- Removed commented-out view entries in the manifest for better readability.
- Simplified date handling in the EstatePropertyOffer model by eliminating redundant checks.
- Updated action methods to directly set status values, enhancing code efficiency.
- Refactored property iteration in the EstateProperty model for consistency and clarity.
